src/lib/ext-activitylog.js

```diff
@@ -0,0 +1,120 @@
+    this.logTableWrapper = document.querySelector('.log-table');
+    this.tableBody = document.querySelector('table tbody');
+    this.notice = document.querySelector('.notice');
+    this.closeBtn = document.querySelector('.close');
+  }
+
+  async init() {
+    const existingLogs = await this.getExistingLogs();
+
+    if (existingLogs.length) {
+      this.addNewLogs(existingLogs);
+    }
+
+    this.saveLogBtn.addEventListener('click', this);
+    this.closeBtn.addEventListener('click', this);
+    this.tableBody.addEventListener('click', this);
+
+    browser.runtime.onMessage.addListener((message) => {
+      const { requestTo, requestType } = message;
+
+      if (requestTo !== 'activity-log') {
+        return;
+      }
+
+      if (requestType === 'appendLogs') {
+        this.addNewLogs([message.log]);
+      } else {
+        throw new Error(`wrong request type found ${requestType}`);
+      }
+    });
+  }
+
+  async getExistingLogs() {
+    const { existingLogs } = await browser.runtime.sendMessage({
+      requestType: 'sendAllLogs',
+      requestTo: 'ext-monitor',
+    });
+    return existingLogs;
+  }
+
+  addNewLogs(logs) {
+    for (const log of logs) {
+      const newRow = this.tableBody.insertRow(-1);
+      newRow.insertCell(0).textContent = log.id;
+      newRow.insertCell(1).textContent = log.timeStamp;
+      newRow.insertCell(2).textContent = log.type;
+      newRow.insertCell(3).textContent = log.name;
+      newRow.insertCell(4).textContent = log.viewType || 'undefined';
+      newRow._log = log;
+
+      this.logs.push(log);
+    }
+  }
+
+  setError(errorMessage) {
+    if (errorMessage) {
+      this.notice.textContent = errorMessage;
+      this.notice.classList.add('failure');
+    } else {
+      this.notice.textContent = '';
+      this.notice.classList.remove('failure');
+    }
+  }
+
+  handleEvent(event) {
+    if (event.type === 'click') {
+      const logDetails = event.target.closest('tr')?._log;
+
+      if (logDetails) {
+        this.openDetailSidebar(logDetails);
+        return;
+      }
+
+      switch (event.target) {
+        case this.saveLogBtn:
+          this.saveLogs();
+          break;
+        case this.closeBtn:
+          this.closeDetailSidebar();
+          break;
+        default:
+          throw new Error(`unexpected click event on ${event.target}`);
+      }
+    } else {
+      throw new Error(`wrong event type found ${event.type}`);
+    }
+  }
+
+  openDetailSidebar(logDetails) {
+    const logString = JSON.stringify(logDetails);
+    this.showLogDetails.textContent = logString;
+    this.showLogDetailWrapper.style.display = 'block';
+    this.logTableWrapper.classList.add('width-60');
+  }
+
+  closeDetailSidebar() {
+    this.showLogDetailWrapper.style.display = 'none';
+    this.logTableWrapper.classList.remove('width-60');
+  }
+
+  async saveLogs() {
+    try {
+      await save.saveAsJSON(this.logs);
+      this.setError(false);
```

Comment:
I put `false` here for readability purpose.

Reply:
using false to clear the event makes the expected type signature of the `setError` method a bit odd:

- by passing a string or null/undefined, we can say that the type signature for `setErrot` would be something like `setError(errorMessage: string|void): void` (gets a string or nothing and return nothing)

- by passing an error message `string` or `false`, then `setError` would have a type signature that would look like `setError(errorMessage: string|bool): void` (gets a string or a boolean and return nothing)

In my opinion, `setError(errorMessage: string|bool): void` is a more reasonable type signature for that function.

Reply:
Currently, it is in `setError(errorMessage: string|bool): void` format, right ? Like `errorMessage: false` if I am sending `false` as argument. And `errorMessage: string` when I am sending the `error.message` as argument.